### PR TITLE
OCPBUGS-86480: E2E:  Refactor OVS affinity tests into shared helpers and add new workload-partitioning and control-plane reboot test cases.

### PR DESCRIPTION
**Pull request by mrniranjan**
This PR has 2 commits
1. First commit refactors some of the existing code  in terms of readability etc 
2. Add new test cases related related to workload partitioning 

Code was generated using AI-Assisted:
AIA Human-AI blend, Content edits, Human-initiated, claude-4.6-opus-high-thinking

## Summary by CodeRabbit

* **Tests**
  * Enhanced performance-profile tests to gate CPU-affinity expectations based on Workload Partitioning (enabled/disabled).
  * Updated OVN/OVS affinity checks to validate cpuset behavior, including ovs-vswitchd thread affinity and multi-pod scenarios.
  * Rewrote dynamic OVS pinning tests to cover GU pod pinning, exclusion from OVS affinity, and persistence across pod deletion and node/control-plane reboots.
* **Chores**
  * Added a cluster utility to detect Workload Partitioning status.

### CODE REVIEW

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 1]
```diff
@@ -78,6 +81,14 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
 
 		ovsSystemdServices = ovsSystemdServicesOnOvsSlice(ctx, workerRTNode)
 
+		isWorkloadPartitioningEnabled, err = cluster.IsWorkloadPartitioningEnabled(ctx)
+		Expect(err).ToNot(HaveOccurred(), "Unable to check if workload partitioning is enabled")
+		testlog.Infof("Workload partitioning enabled: %v", isWorkloadPartitioningEnabled)
+
+		Expect(profile.Spec.CPU.Reserved).NotTo(BeNil())
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: nit: `ToNot`

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 2]
```diff
@@ -128,7 +139,13 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
 					cpus := testutils.ToString(out)
 					containerCpuset, err := cpuset.Parse(cpus)
 					Expect(err).ToNot(HaveOccurred())
-					Expect(containerCpuset).To(Equal(onlineCPUSet), "Burstable pod containers cpuset.cpus do not match total online cpus")
+					if isWorkloadPartitioningEnabled {
+						Expect(containerCpuset).To(Equal(reservedCPUSet),
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Use the built-in `containerCpuset.Equals()` function

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 3]
```diff
@@ -128,7 +139,13 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
...
+							"Under workload partitioning, OVN pod cpuset.cpus should match reserved cpus, got %s expected %s", containerCpuset, reservedCPUSet)
+					} else {
+						Expect(containerCpuset).To(Equal(onlineCPUSet),
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: ditto

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 4]
```diff
@@ -265,448 +282,194 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
...
 				defer func() {
-					// delete deployment
-					testlog.Infof("Deleting Deployment %v", dp.Name)
-					err := testclient.DataPlaneClient.Delete(ctx, dp)
-					Expect(err).ToNot(HaveOccurred())
-					// once deployment is deleted
-					// wait till the ovs process affinity is reverted back
-					pidList, err := ovsPids(ctx, ovsSystemdServices, workerRTNode)
-					Expect(err).ToNot(HaveOccurred())
+					testlog.Infof("Deleting Deployment %s", dp.Name)
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: maybe add NS as well for better logging experience

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 5]
```diff
@@ -265,448 +282,194 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
...
 				}
 
-				testlog.Info("Rebooting the node")
-				rebootCmd := "chroot /rootfs systemctl reboot"
+				waitForDeploymentReady(ctx, dp, dpListOpts, 2)
+
+				By("Verifying affinities before reboot")
+				collectAndVerify("before-reboot")
+
+				By("Rebooting the worker node")
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: you may consolidate with the line below.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 6]
```diff
@@ -910,3 +673,136 @@ func ovsSwitchdThreadAffinity(ctx context.Context, workerRTNode *corev1.Node) ([
 	threadAffinity := strings.Split(string(out), "\n")
 	return threadAffinity, nil
 }
+
+// expectedOvsAffinity computes the expected OVS process CPU affinity set.
+// Formula: (Allocatable - GU_Pinned) ∪ Reserved
+// where Allocatable = Online - Reserved (i.e., the isolated CPUs).
+func expectedOvsAffinity(onlineCPUs, reservedCPUs, guPodCPUs cpuset.CPUSet) cpuset.CPUSet {
+	isolated := onlineCPUs.Difference(reservedCPUs)
+	return isolated.Difference(guPodCPUs).Union(reservedCPUs)
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: or simply:
` return onlineCPUs.Difference(guPodCPUs)`

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 7]
```diff
@@ -910,3 +673,136 @@ func ovsSwitchdThreadAffinity(ctx context.Context, workerRTNode *corev1.Node) ([
...
+}
+
+// getOvnContainerAffinity returns the CPU affinity of the first OVN container
+// on the given node, waiting 30s for affinity to stabilise.
+func getOvnContainerAffinity(ctx context.Context, node *corev1.Node) cpuset.CPUSet {
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go:940]: add `GinkgoHelper()` in all test helpers that are throwing assertions outside the `It()` containers

**Review by Tal-or — COMMENTED** [review]
Good work, couple of comments inside.

On the u/s the cluster are not configured with WP, are we going to test cluster with WP configured d/s?

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 5 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 6 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 7 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go:940]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 8]
```diff
@@ -910,3 +928,144 @@ func ovsSwitchdThreadAffinity(ctx context.Context, workerRTNode *corev1.Node) ([
...
+			return false
+		}
+		if len(podList.Items) == 0 {
+			return false
+		}
+		for _, s := range podList.Items[0].Status.ContainerStatuses {
+			if !s.Ready {
+				return false
+			}
+		}
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go:1073]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 4 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 3 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 2 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 1 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Addressed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 6 above)
Comment (MarSik) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: This is probably valid for now, but we will have to change it again once @Tal-or reviews the dedicated cpus PR. I wonder if we could join the systemd cpu affinity + Allocatable - GU_Pinned and make it future proof.

File: test/e2e/performanceprofile/functests/utils/cluster/cluster.go [hunk 9]
```diff
@@ -47,3 +48,19 @@ func IsControlPlaneSchedulable(ctx context.Context) (bool, error) {
 	}
 	return schedulerInfo.Spec.MastersSchedulable, nil
 }
+
+// IsWorkloadPartitioningEnabled checks whether CPU partitioning is enabled
+// cluster-wide by querying the Infrastructure resource's CPUPartitioning status.
```
Comment (MarSik) [inline test/e2e/performanceprofile/functests/utils/cluster/cluster.go:43]: WP might be enabled, but not effective when the PerformanceProfile is not applied. That is probably not important for our CI, but it is a corner case (E2E upgrade might be using paused MCPs without PP applied).

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 10]
```diff
@@ -390,6 +399,83 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
...
+				nodes.WaitForReadyOrFail("Reboot", cpNode.Name, 10*time.Minute, 30*time.Second)
+
+				By("Waiting for OVN pod to be ready on the control plane node")
+				Eventually(func() error {
+					_, err := ovnCnfNodePod(ctx, cpNode)
+					return err
+				}, 5*time.Minute, 10*time.Second).Should(Succeed(),
+					"OVN pod did not become ready on control plane node after reboot")
+
+				By("Verify OVS affinity after reboot is not restricted to reserved CPUs")
```
Comment (MarSik) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Is this giving enought time to ovn-k to change the cpu affinity? In other words, will it be racy on say slow clusters?

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 11]
```diff
@@ -456,23 +460,13 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
 				}, 5*time.Minute, 10*time.Second).Should(Succeed(),
 					"OVN pod did not become ready on control plane node after reboot")
 
-				By("Verify OVS affinity after reboot is not restricted to reserved CPUs")
+				By("Verify OVS affinity after reboot spans all online CPUs")
```
Comment (MarSik) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: This will not be true after dedicated cpus are introduced as I mentioned before.

**Comment by Tal-or** [comment]
@MarSik I think it would be easier to create a follow up PR once dedicated CPUs are in.
I'll add a note to myself, to revise those tests in the future.
[CNF-23698](https://redhat.atlassian.net/browse/CNF-23698 )

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 6 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: @MarSik Can i use  OvsSliceCpuset - GU_Pinned ?
 The OVS slice cpuset (read from cgroup) reflects the actual kernel-enforced CPU constraint. Subtracting GU-pinned CPUs yields the correct expected affinity . I.e we read cpuset.cpus.effective from the ovs.slice    and use that to subtract the GU_Pinned ?

Comment (MarSik) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: Now it does, but it won't in the future. @Tal-or can show you the details, but OVS should be checked to run on PerfProfile.spec.cpus.reserved + PerfProfile.spec.cpus.isolated - GU_pinned. This way it will be future proof.

Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: @MarSik fixed in the latest commit.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 10 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: I didn't notice any issues when i ran on compact clusters (BM)

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 11 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/utils/cluster/cluster.go (hunk 9 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cluster/cluster.go:43]: @MarSik we don't run our tests in CI without Performance profile. In fact without performance profile tests would not even start.

**Review by MarSik — APPROVED** [review]
/lgtm
/hold for Talor to take a look

File: test/e2e/performanceprofile/functests/utils/cluster/cluster.go [hunk 12]
```diff
@@ -47,3 +48,19 @@ func IsControlPlaneSchedulable(ctx context.Context) (bool, error) {
...
+func IsWorkloadPartitioningEnabled(ctx context.Context) (bool, error) {
+	cfg, err := config.GetConfig()
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/utils/cluster/cluster.go]: Why not passing the existing runtime client? initializing new client every time this function is being called is not a good practice.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 13]
```diff
@@ -910,3 +916,160 @@ func ovsSwitchdThreadAffinity(ctx context.Context, workerRTNode *corev1.Node) ([
 	threadAffinity := strings.Split(string(out), "\n")
 	return threadAffinity, nil
 }
+
+// expectedOvsAffinity computes the expected OVS process CPU affinity set.
+// Formula: (reserved + isolated) - GU_Pinned
+// reserved+isolated is the current profile-derived baseline for OVS. Subtracting
+// GU-pinned CPUs yields the expected affinity in workload partitioning mode.
+func expectedOvsAffinity(reservedCPUs, isolatedCPUs, guPodCPUs cpuset.CPUSet) cpuset.CPUSet {
+	return reservedCPUs.Union(isolatedCPUs).Difference(guPodCPUs)
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go:935]: Yes. that would be always correct, even with dedicated CPUs feature

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 14]
```diff
@@ -265,473 +283,459 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
...
+			It("[test_id:89063] Verify OVS affinity is wider than OVN pod under workload partitioning", func() {
+				By("Get OVN container affinity")
+				ovnPod, err := ovnCnfNodePod(ctx, workerRTNode)
+				Expect(err).ToNot(HaveOccurred(), "Unable to get ovnPod")
+				ovnContainerids, err := ovnPodContainers(&ovnPod)
+				Expect(err).ToNot(HaveOccurred())
+				containerPid, err := nodes.ContainerPid(ctx, workerRTNode, ovnContainerids[0])
+				Expect(err).ToNot(HaveOccurred())
+
+				time.Sleep(30 * time.Second)
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: I'm assuming that we're waiting for some state to reconcile. 
Instead of using `sleep` we may use `Eventually` which allow more flexible timeout and return early if the condition was satisfied.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 15]
```diff
@@ -265,473 +283,459 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, Label(string(lab
...
-				for pid, cpumask := range pidToCPUs {
-					testlog.Infof("OVS service pid %s is using cpus %s", pid, cpumask)
-					Expect(ctnCpuset.Equals(cpumask)).To(BeTrue(), "affinity of ovn kube node pods(%s) do not match with ovservices pid %s (%s)", ctnCpuset, pid, cpumask)
+				By("Collecting OVN container and OVS process affinities")
+				ovnAffinity := getOvnContainerAffinity(ctx, workerRTNode)
+				ovsAffinities := getOvsAffinities(ctx, ovsSystemdServices, workerRTNode)
+
+				By("Verifying OVS affinity matches expected")
+				if isWorkloadPartitioningEnabled {
+					verifyOvsMatchesExpectedWP(ovnAffinity, ovsAffinities,
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: I think it would be simpler to merge both `verifyOvsMatchesExpectedWP` and `verifyOvsMatchesExpected` together and pass `isWorkloadPartitioningEnabled` as an argument to the merged function, avoiding the `if/else` flow in several places.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 16]
```diff
@@ -910,3 +916,160 @@ func ovsSwitchdThreadAffinity(ctx context.Context, workerRTNode *corev1.Node) ([
...
+	GinkgoHelper()
+	for pid, mask := range ovsAffinities {
+		testlog.Infof("OVS pid %s affinity: %s (expected: %s)", pid, mask, expected)
+		Expect(mask).To(Equal(expected))
+	}
+}
+
+// verifyOvsMatchesExpected verifies non-workload-partitioning behavior where
+// OVS process affinity should match the OVN container affinity.
+func verifyOvsMatchesExpected(ovnAffinity cpuset.CPUSet,
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go:977]: you may also check the OVN affinity which should be reserved+isolated. 
this check will protect us in the future when the dedicated CPU feature will introduce. because we don't want OVN pod (or any other pod in this case) to span across the dedicated CPUs.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go [hunk 17]
```diff
@@ -910,3 +916,160 @@ func ovsSwitchdThreadAffinity(ctx context.Context, workerRTNode *corev1.Node) ([
...
+	ovsAffinities map[string]cpuset.CPUSet,
+	reservedCPUs, isolatedCPUs, guPodCPUs cpuset.CPUSet) {
+	GinkgoHelper()
+	Expect(ovnAffinity).To(Equal(reservedCPUs),
+		"Under WP, OVN container should be restricted to reserved cpus")
+	expected := expectedOvsAffinity(reservedCPUs, isolatedCPUs, guPodCPUs)
+	verifyOvsAffinity(ovsAffinities, expected)
+}
+
+func parseProfileCPUSets(profile *performancev2.PerformanceProfile) (cpuset.CPUSet, cpuset.CPUSet) {
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: I would return a `map` and let this function try to parse all types of CPUs we have:
`reserved`, `isolated`, `shared`, `offlined`.

if some are not declared the value would be an empty set.

**Review by Tal-or — COMMENTED** [review]
looking good overall. 
left some comments.

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 17 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: fixed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 16 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go:977]: fixed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 15 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: fixed in latest commit

File: test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go (hunk 14 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go]: fixed in latest commit

File: test/e2e/performanceprofile/functests/utils/cluster/cluster.go (hunk 12 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cluster/cluster.go]: fixed in latest commit

**Review by Tal-or — APPROVED** [review]
/approve
/lgtm
/hold cancel

Thank you!
